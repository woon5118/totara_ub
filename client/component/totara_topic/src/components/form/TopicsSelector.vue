<!--
  This file is part of Totara Enterprise Extensions.

  Copyright (C) 2020 onwards Totara Learning Solutions LTD

  Totara Enterprise Extensions is provided only to Totara
  Learning Solutions LTD's customers and partners, pursuant to
  the terms and conditions of a separate agreement with Totara
  Learning Solutions LTD or its affiliate.

  If you do not have an agreement with Totara Learning Solutions
  LTD, you may not access, use, modify, or distribute this software.
  Please contact [licensing@totaralearning.com] for more information.

  @author Kian Nguyen <kian.nguyen@totaralearning.com>
  @module totara_topic
-->
<template>
  <TagList
    class="tui-topicsSelector"
    :filter="searchTerm"
    :items="topics"
    :tags="pickedTopics"
    :input-placeholder="inputPlaceholder"
    @filter="findTopics"
    @select="selectTopic"
    @remove="removeTopic"
  >
    <template v-slot:item="{ item: { value } }">
      {{ value }}
    </template>
  </TagList>
</template>

<script>
import TagList from 'tui/components/tag/TagList';

// GraphQL queries
import findTopics from 'totara_topic/graphql/find_topics';

export default {
  components: {
    TagList,
  },

  props: {
    selectedTopics: {
      type: Array,
      default() {
        return [];
      },

      validator(prop) {
        let items = Array.prototype.filter.call(prop, item => {
          return !('value' in item) || !('id' in item);
        });

        return 0 === items.length;
      },
    },

    inputPlaceholder: String,
  },

  apollo: {
    topics: {
      query: findTopics,
      variables() {
        return {
          search: this.searchTerm,
          exclude: this.selectedTopicIds,
        };
      },
    },
  },

  data() {
    return {
      searchTerm: '',
      topics: [],
    };
  },

  computed: {
    selectedTopicIds() {
      if (0 === this.selectedTopics.length) {
        return [];
      }

      return Array.prototype.map.call(this.selectedTopics, ({ id }) => id);
    },

    pickedTopics() {
      return Array.prototype.map.call(this.selectedTopics, ({ value, id }) => {
        return {
          id: id,
          text: value,
        };
      });
    },
  },

  methods: {
    /**
     *
     * @param {String} term
     */
    findTopics(term) {
      this.searchTerm = term;
      this.$apollo.queries.topics.refetch();
    },

    /**
     *
     * @param {Number}  id
     * @param {String}  value
     */
    selectTopic({ id, value }) {
      const selectedTopics = Array.prototype.concat.call(this.selectedTopics, {
        id,
        value,
      });

      this.$emit('change', selectedTopics);

      // Reset search term and refetch the query.
      this.searchTerm = '';
      this.$apollo.queries.topics.refetch();
    },

    /**
     *
     * @param {Number} id
     */
    removeTopic({ id }) {
      const selectedTopics = Array.prototype.filter.call(
        this.selectedTopics,
        item => {
          return item.id !== id;
        }
      );

      // Need to update the list again.
      this.$emit('change', selectedTopics);
      this.$apollo.queries.topics.refetch();
    },
  },
};
</script>

<style lang="scss">
.tui-topicsSelector {
  width: 100%;
}
</style>
