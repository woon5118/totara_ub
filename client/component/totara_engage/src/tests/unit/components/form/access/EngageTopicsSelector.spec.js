/**
 * This file is part of Totara Enterprise Extensions.
 *
 * Copyright (C) 2020 onwards Totara Learning Solutions LTD
 *
 * Totara Enterprise Extensions is provided only to Totara
 * Learning Solutions LTD's customers and partners, pursuant to
 * the terms and conditions of a separate agreement with Totara
 * Learning Solutions LTD or its affiliate.
 *
 * If you do not have an agreement with Totara Learning Solutions
 * LTD, you may not access, use, modify, or distribute this software.
 * Please contact [licensing@totaralearning.com] for more information.
 *
 * @author Alvin Smith <alvin.smith@totaralearning.com>
 * @module totara_engage
 */

import TopicsSelector from 'totara_engage/components/form/access/EngageTopicsSelector';
import { shallowMount } from '@vue/test-utils';

describe('totara_engage/components/form/access/EngageTopicsSelector.vue', function() {
  it('Checks snapshot', () => {
    let wrapper = shallowMount(TopicsSelector, {
      mocks: {
        $str(id, component) {
          return `${id}, ${component}`;
        },
        $id: x => 'id-' + (x || 'topic'),
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
