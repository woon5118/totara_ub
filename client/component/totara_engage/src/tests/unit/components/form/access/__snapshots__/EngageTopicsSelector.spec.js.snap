// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`totara_engage/components/form/access/EngageTopicsSelector.vue Checks snapshot 1`] = `
<div
  class="tui-engageTopicsSelector"
>
  <label-stub
    class="tui-engageTopicsSelector__label"
    forid="id-topic"
    label="assigntopics, totara_engage"
  />
   
  <topicsselector-stub
    id="id-topic"
    inputplaceholder="enter_topics, totara_engage"
    selectedtopics=""
  />
</div>
`;
